examples/petsc/dmswarm.c

```diff
@@ -0,0 +1,412 @@
+  for (PetscInt cell = cell_start; cell < cell_end; cell++) {
+    PetscReal *coords_cell_points_true, *coords_cell_points_ref;
+    PetscInt  *cell_points;
+    PetscInt   num_points_in_cell;
+
+    PetscCall(DMSwarmSortGetPointsPerCell(dm_swarm, cell, &num_points_in_cell, &cell_points));
+    PetscCall(DMGetWorkArray(dm_mesh, num_points_in_cell * dim, MPIU_REAL, &coords_cell_points_true));
+    PetscCall(DMGetWorkArray(dm_mesh, num_points_in_cell * dim, MPIU_REAL, &coords_cell_points_ref));
+    // -- Reference coordinates for swarm points in background mesh element
+    for (PetscInt p = 0; p < num_points_in_cell; p++) {
+      point_cell_numbers[cell_points[p]] = cell;
+      for (PetscInt d = 0; d < dim; d++) coords_cell_points_true[p * dim + d] = coords_points_true[cell_points[p] * dim + d];
+    }
+    PetscCall(DMPlexCoordinatesToReference(dm_mesh, cell, num_points_in_cell, coords_cell_points_true, coords_cell_points_ref));
+    for (PetscInt p = 0; p < num_points_in_cell; p++) {
+      for (PetscInt d = 0; d < dim; d++) coords_points_ref[cell_points[p] * dim + d] = coords_cell_points_ref[p * dim + d];
+    }
+
+    // -- Cleanup
+    PetscCall(DMRestoreWorkArray(dm_mesh, num_points_in_cell * dim, MPIU_REAL, &coords_cell_points_true));
+    PetscCall(DMRestoreWorkArray(dm_mesh, num_points_in_cell * dim, MPIU_REAL, &coords_cell_points_ref));
+    PetscCall(PetscFree(cell_points));
+  }
+
+  // Cleanup
+  PetscCall(DMSwarmRestoreField(dm_swarm, "DMSwarmPIC_coor", NULL, NULL, (void **)&coords_points_ref));
+  PetscCall(VecRestoreArrayRead(*ref_coords, &coords_points_true));
+  PetscCall(DMSwarmSortRestoreAccess(dm_swarm));
+
+  // Create index set
+  PetscCall(ISCreateGeneral(PETSC_COMM_SELF, num_points_local, point_cell_numbers, PETSC_OWN_POINTER, is_points));
+
+  PetscFunctionReturn(PETSC_SUCCESS);
+}
+
+int main(int argc, char **argv) {
+  MPI_Comm            comm;
+  DM                  dm_mesh, dm_swarm;
+  Vec                 U_mesh;
+  PetscInt            dim = 3, num_comp = 1, num_points = 10, geometry_order = 1, mesh_order = 2, q_order = 3;
+  Ceed                ceed;
+  CeedBasis           basis_mesh_u;
+  CeedElemRestriction restriction_mesh_u;
+
+  PetscCall(PetscInitialize(&argc, &argv, NULL, help));
+  comm = PETSC_COMM_WORLD;
+  CeedInit("/cpu/self", &ceed);
+
+  // Create background mesh
+  PetscCall(DMCreate(comm, &dm_mesh));
+  PetscCall(DMSetType(dm_mesh, DMPLEX));
+  PetscCall(DMSetFromOptions(dm_mesh));
+
+  // -- Mesh FE space
+  {
+    PetscFE fe;
+
+    PetscCall(DMGetDimension(dm_mesh, &dim));
+    PetscCall(PetscFECreateLagrange(comm, dim, num_comp, PETSC_FALSE, mesh_order, q_order, &fe));
+    PetscCall(DMAddField(dm_mesh, NULL, (PetscObject)fe));
+    PetscCall(PetscFEDestroy(&fe));
+  }
+  PetscCall(DMCreateDS(dm_mesh));
+
+  // -- Coordinate FE space
+  {
+    PetscFE fe_coord;
+
+    PetscCall(PetscFECreateLagrange(comm, dim, num_comp, PETSC_FALSE, geometry_order, q_order, &fe_coord));
```

Comment:
Small note, I think this should be:
```suggestion
    PetscCall(PetscFECreateLagrange(comm, dim, dim, PETSC_FALSE, geometry_order, q_order, &fe_coord));
```